src/python/pants/goal/migrate_call_by_name.py

```diff
@@ -3,35 +3,655 @@
+        """Collect all names in the file, so we can avoid shadowing them with the new imports."""
+        self.names.add(node.id)
+
+    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef):
+        """This function will determine if the node is a rule, if there are any Get(...) calls of
+        interest, and then call out to the syntax mapper to create the new syntax and imports for
+        the replacement."""
+
+        # Ensure we collect all names in this function, as well
+        names = [n.id for n in ast.walk(node) if isinstance(n, ast.Name)]
+        self.names.update(names)
+
+        if not self._should_visit_node(node.decorator_list):
+            return
+
+        for child in node.body:
+            if call := self._maybe_replaceable_call(child):
+                if replacement := self.syntax_mapper.map_get_to_new_syntax(
+                    call, self.filename, calling_func=node.name
+                ):
+                    self.replacements.append(replacement)
+
+            for call in self._maybe_replaceable_multiget(child):
+                if replacement := self.syntax_mapper.map_get_to_new_syntax(
+                    call, self.filename, calling_func=node.name
+                ):
+                    replacement.is_argument = True
+                    self.replacements.append(replacement)
+
+    def _should_visit_node(self, decorator_list: list[ast.expr]) -> bool:
+        """Only interested in async functions with the @rule(...) or @goal_rule(...) decorator."""
+        for decorator in decorator_list:
+            if isinstance(decorator, ast.Name) and decorator.id in ["rule", "goal_rule"]:
+                # Accounts for isolated decorators e.g. "@rule"
+                return True
+
+            if (
+                isinstance(decorator, ast.Call)
+                and isinstance(decorator.func, ast.Name)
+                and decorator.func.id in ["rule", "goal_rule"]
+            ):
+                # Accounts for decorators with params e.g. "@rule(desc=..., level=...)"
+                return True
+
+        return False
+
+    def _maybe_replaceable_call(self, statement: ast.stmt) -> ast.Call | None:
+        """Looks for the following forms of Get that we want to replace:
+
+        - bar_get = Get(...)
+        - bar = await Get(...)
+        This function is pretty gross, but a lot of it is to satisfy the type checker and
+        to not need to try/catch everywhere.
+        """
+        if (
+            isinstance(statement, ast.Assign)
+            and (
+                isinstance((call_node := statement.value), ast.Call)
+                or (
+                    isinstance((await_node := statement.value), ast.Await)
+                    and isinstance((call_node := await_node.value), ast.Call)
+                )
+            )
+            and isinstance(call_node.func, ast.Name)
+            and call_node.func.id == "Get"
+        ):
+            return call_node
+        return None
```

Comment:
AFAIK, any call to `Get(..)` should be replaceable to `function_call(..)`: the only difference should be that before replacement you get out a `Get[T]`, and after replacement you get out an `Awaitable[T]`... so it might break some type checking I suppose. But maybe it would leave you happier with this code?

Reply:
Yeah, this is more about satisfying the type checker, as if the call is awaitable, we access a different part of the AST vs if it's just straight callable. This will be pulled out to a utility to at least make the primary code more linearly readable.